### PR TITLE
Add argument inference_grad_mode to Trainer

### LINKED ISSUE #14497: Need an option for evaluation not in the inference_mode.
## 🚀 Feature

An option for the trainer run evaluation not in the inference_mode.

### Motivation

In some scenerios, gradients are used even when evaluating the model. The recent PR (https://github.com/Lightning-AI/lightning/pull/12715) adds supports for the new PyTorch API `inference_mode` and enables it by default. However, unlike `torch.no_grad` that one can simply set `torch.enable_grad` to overwrite it locally, `inference_mode` do not provide a convenient way to quit it locally. This is because all tensors are marked and forbidden for furthur gradient-related usage. People have to clone tensors to make code working. Therefore, I suggest that pytorch-lightning provide a way to let users to choose whether to use the inference mode.

### Pitch

Add an option to the Trainer class that allows users to decide whether to use the inference mode or not.

### Additional context

From the pytorch's document https://pytorch.org/cppdocs/notes/inference_mode.html:
> Inference tensors:
> ...
> are immutable outside InferenceMode. So an error will be raised if you try to: - mutate their data outside InferenceMode. - mutate them into requires_grad=True outside InferenceMode. To work around you can make a clone outside InferenceMode to get a normal tensor before mutating.

cc @justusschock @kaushikb11 @awaelchli @borda @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by LouChao98**
## What does this PR do?

Fixes #14497. Add `inference_grad_mode` to `Trainer` to enable users to control the grad mode used when inference.

### Does your PR introduce any breaking changes? If yes, please list them.

Expected to be none.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by akihironitta — COMMENTED** [review]
IMO, it would make more sense to have the argument via LightningModule's attribute just like `self.automatic_optimization=True|False` because whether one should use the inference mode or not only depends on each validation/test implementation, i.e., on whether they use grads or not. Or, is there any other case that users would want to switch the argument?

**Comment by rohitgr7** [comment]
> 

I am fine with either since it's completely related to LM but trainer is the one that calls it.

File: src/pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -161,6 +161,7 @@ def __init__(
         limit_test_batches: Optional[Union[int, float]] = None,
         limit_predict_batches: Optional[Union[int, float]] = None,
         val_check_interval: Optional[Union[int, float]] = None,
+        inference_grad_mode: Optional[str] = None,
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:164]: just in case it stays in Trainer, please at it at the end.

File: src/pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -529,6 +533,13 @@ def __init__(
         )
         self.track_grad_norm: float = float(track_grad_norm)
 
+        if inference_grad_mode not in {"enable_grad", "no_grad", "inference_mode", None}:
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:536]: I think there should be just 2. or just `use_inference_mode=True|False`.

**Comment by rschireman** [comment]
Just saw that this got marked stale, I would love to help the development since this feature would be really useful

**Comment by rohitgr7** [comment]
@rschireman go ahead!

**Comment by akihironitta** [comment]
Hi @LouChao98, as #15034 has got merged already, would you mind closing this PR? Thank you for your contribution!
